src/common/ceph_crypto.h

```diff
@@ -20,61 +21,148 @@
-	Restart();
-      }
-      ~Digest () {
-	PK11_DestroyContext(ctx, PR_TRUE);
-      }
-      void Restart() {
-	SECStatus s;
-	s = PK11_DigestBegin(ctx);
-	ceph_assert_always(s == SECSuccess);
-      }
-      void Update (const unsigned char *input, size_t length) {
-        if (length) {
+      Digest () {};
+      virtual ~Digest () {}
+      virtual void Restart() = 0;
+      virtual void Update (const unsigned char *input, size_t length) = 0;
+      virtual void Final (unsigned char *digest) = 0;
+    };
+  }
+}
+
+#ifdef USE_NSS
+namespace ceph {
+  namespace crypto {
+    namespace nss {
+      class NSSDigest : public Digest {
+      private:
+        PK11Context *ctx;
+        size_t digest_size;
+      public:
+        NSSDigest (SECOidTag _type, size_t _digest_size)
+	  : Digest()
+	  , digest_size(_digest_size) {
+	  ctx = PK11_CreateDigestContext(_type);
+	  ceph_assert_always(ctx);
+	  Restart();
+        }
+        virtual ~NSSDigest () {
+	  PK11_DestroyContext(ctx, PR_TRUE);
+	}
+	virtual void Restart() {
 	  SECStatus s;
-	  s = PK11_DigestOp(ctx, input, length);
+	  s = PK11_DigestBegin(ctx);
 	  ceph_assert_always(s == SECSuccess);
-        }
-      }
-      void Final (unsigned char *digest) {
-	SECStatus s;
-	unsigned int dummy;
-	s = PK11_DigestFinal(ctx, digest, &dummy, digest_size);
-	ceph_assert_always(s == SECSuccess);
-	ceph_assert_always(dummy == digest_size);
-	Restart();
-      }
-    };
-    class MD5 : public Digest {
-    public:
-      MD5 () : Digest(SEC_OID_MD5, CEPH_CRYPTO_MD5_DIGESTSIZE) { }
-    };
+	}
+	virtual void Update (const unsigned char *input, size_t length) {
+	  if (length) {
+	    SECStatus s;
+	    s = PK11_DigestOp(ctx, input, length);
+	    ceph_assert_always(s == SECSuccess);
+	  }
+	}
+	virtual void Final (unsigned char *digest) {
+	  SECStatus s;
+	  unsigned int dummy;
+	  s = PK11_DigestFinal(ctx, digest, &dummy, digest_size);
+	  ceph_assert_always(s == SECSuccess);
+	  ceph_assert_always(dummy == digest_size);
+	  Restart();
+	}
+      };
 
-    class SHA1 : public Digest {
-    public:
-      SHA1 () : Digest(SEC_OID_SHA1, CEPH_CRYPTO_SHA1_DIGESTSIZE) { }
-    };
+      class MD5 : public NSSDigest {
+      public:
+	MD5 () : NSSDigest(SEC_OID_MD5, CEPH_CRYPTO_MD5_DIGESTSIZE) { }
+      };
 
-    class SHA256 : public Digest {
-    public:
-      SHA256 () : Digest(SEC_OID_SHA256, CEPH_CRYPTO_SHA256_DIGESTSIZE) { }
-    };
+      class SHA1 : public NSSDigest {
+      public:
+        SHA1 () : NSSDigest(SEC_OID_SHA1, CEPH_CRYPTO_SHA1_DIGESTSIZE) { }
+      };
+
+      class SHA256 : public NSSDigest {
+      public:
+        SHA256 () : NSSDigest(SEC_OID_SHA256, CEPH_CRYPTO_SHA256_DIGESTSIZE) { }
+      };
+    }
+  }
+}
+#endif /*USE_NSS*/
+
+#ifdef USE_OPENSSL
+namespace ceph {
+  namespace crypto {
+    namespace ssl {
+      class OpenSSLDigest : public Digest {
+      private:
+	EVP_MD_CTX *mpContext;
+	const EVP_MD *mpType;
+      public:
+	OpenSSLDigest (SECOidTag _type);
```

Comment:
SECOidTag is a nss type, right? can we find a way to avoid depending on this for the openssl implementations?

Reply:
Have removed.